src/repr/src/timestamp.rs

```diff
@@ -85,6 +85,30 @@ impl Timestamp {
         }
     }
 
+    /// Creates a [`Timestamp`] from a [`Duration`]
+    /// 
+    /// # Panics
+    /// * If the provided [`Duration`], when represented in milliseconds, overflows a `u64`
+    /// 
+    /// ```should_panic
+    /// use mz_repr::Timestamp;
+    /// use std::time::Duration;
+    /// 
+    /// let d = Duration::from_secs(u64::MAX);
+    /// let t = Timestamp::from_duration(d);
+    /// ```
+    pub const fn from_duration(duration: Duration) -> Self {
+        let millis = duration.as_millis();
+        assert!(
+            millis <= u64::MAX as u128,
+            "overflow when creating Timestamp from Duration!"
+        );
+
+        Self {
+            internal: millis as u64,
+        }
```

Comment:
Thanks for the PR!

Our linter will likely complain here about using `as`, since we prefer to avoid `as`.

I think we should add a `const` version of `try_cast_from` that panics when the cast fails instead of returning `None`, here:
https://github.com/MaterializeInc/materialize/blob/efbc8dfc6ae0ee29a1c2d6c4b81119df4235fa2a/src/ore/src/cast.rs#L194-L212 

It would likely be very similar to the non-fallible version here:
https://github.com/MaterializeInc/materialize/blob/efbc8dfc6ae0ee29a1c2d6c4b81119df4235fa2a/src/ore/src/cast.rs#L37-L57

Reply:
FWIW I'm not sure a macro here would work, since casting a non-primitive type to another non-primitive, probably would require different implementations? What are your thoughts?

Reply:
We'll let @mjibson and @umanwizard determine if it's worth implementing this since they are more familiar with this code. However, if we go forward then I was thinking that the macro would implement a `const` conversion from `u128` to `u64`. Then you can call the generated const function in `from_duration` instead of using `as`.

Reply:
That'd still panic if the conversion fails right? That's my hangup.

Reply:
I think it will still panic if conversion fails, but only at compile time. So the `const` conversions between primitives would be useful if we ever need to do a fallible conversion at compile time, that we know will not fail.

Reply:
Fine with me if it fails at compile time.

Reply:
Wait but could we enforce that it's never called for non-const contexts? I don't want one of us to later call it for user-supplied Durations.

Reply:
There may be a lint, but out of the box it is possible to call this function outside of a const context, in which case the panic will happen at runtime. For example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7dd5c8b6c45b54a692144fb6dbad2a70

Reply:
Here's an example of panicking at compile time: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c87810ed6dffe850d8377e7cf98b7b03.

I don't have a strong opinion on whether adding the const function is worth it or not, and will defer to you and @umanwizard, since you too added this code.